fluent/client/ws/connection_test.go

```diff
@@ -21,212 +22,238 @@ type message struct {
 
 var _ = Describe("Connection", func() {
 	var (
-		fakeConn           *extfakes.FakeConn
-		opts               ws.ConnectionOptions
-		readMsgs           chan message
-		rhCallCt, rmCallCt int32
-		onFail             func(err error)
+		checkClose, checkSvrClose       bool
+		connection, svrConnection       ws.Connection
+		svr                             *httptest.Server
+		opts                            ws.ConnectionOptions
+		svrRcvdMsgs, clientRcvdMsgs     chan message
+		listenErrs                      chan error
+		exitConnState, svrExitConnState ws.ConnState
 	)
 
-	BeforeEach(func() {
-		rmCallCt = 0
-		rhCallCt = 0
-		readMsgs = make(chan message, 1)
-		fakeConn = &extfakes.FakeConn{}
-
-		fakeConn.ReadMessageStub = func() (int, []byte, error) {
-			atomic.AddInt32(&rmCallCt, 1)
-			m, ok := <-readMsgs
-			if !ok {
-				return m.mt, m.msg, errors.New("connection")
-			}
-			return m.mt, m.msg, m.err
-		}
-
-		opts = ws.ConnectionOptions{
+	var makeOpts = func(msgChan chan message, name string) ws.ConnectionOptions {
+		return ws.ConnectionOptions{
+			CloseDeadline: 500 * time.Millisecond,
 			ReadHandler: func(conn ws.Connection, msgType int, p []byte, err error) error {
-				atomic.AddInt32(&rhCallCt, 1)
+				msg := message{
+					mt:  msgType,
+					msg: p,
+					err: err,
+				}
+				msgChan <- msg
+
 				if err != nil {
-					conn.Close()
+					log.Println(name, "ReadHandler received error:", err)
 				}
+
 				return err
 			},
 		}
+	}
 
-		onFail = nil
-	})
+	newHandler := func(svrRcvdMsgs chan message) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			defer GinkgoRecover()
+			svrOpts := makeOpts(svrRcvdMsgs, "server")
 
-	AfterEach(func() {
-		close(readMsgs)
-	})
+			var upgrader websocket.Upgrader
+			wc, _ := upgrader.Upgrade(w, r, nil)
+
+			var err error
+			svrConnection, err = ws.NewConnection(wc, svrOpts)
+			if err != nil {
+				return
+			}
 
-	Describe("NewConnection", func() {
-		It("works", func() {
-			c, err := ws.NewConnection(fakeConn, ws.ConnectionOptions{})
-			Expect(err).ToNot(HaveOccurred())
-			Expect(c).ToNot(BeNil())
-			// TODO: test options are set correctly
+			Expect(svrConnection.Listen()).ToNot(HaveOccurred())
+			log.Println("exit server handler")
 		})
+	}
+
+	BeforeEach(func() {
+		exitConnState = ws.ConnStateCloseReceived | ws.ConnStateCloseSent | ws.ConnStateClosed
+		svrExitConnState = ws.ConnStateCloseReceived | ws.ConnStateCloseSent | ws.ConnStateClosed
+
+		checkSvrClose = true
+		svrRcvdMsgs = make(chan message)
+		svr = httptest.NewServer(newHandler(svrRcvdMsgs))
```

Comment:
Mocking didn't work for these test cases. Some of the cases rely on calls made in the Gorilla websocket package which can only be made with a real network connection.